rln/src/poseidon_tree.rs

```diff
@@ -36,16 +49,101 @@ impl Hasher for PoseidonHash {
     }
 }
 
+// The pmtree Hasher trait used by pmtree Merkle tree
+impl pmtree::Hasher for PoseidonHash {
+    type Fr = Fr;
+
+    fn default_leaf() -> Self::Fr {
+        Fr::from(0)
+    }
+
+    fn serialize(value: Self::Fr) -> Value {
+        fr_to_bytes_le(&value)
+    }
+
+    fn deserialize(value: Value) -> Self::Fr {
+        let (fr, _) = bytes_le_to_fr(&value);
+        //let c = std::io::Cursor::new(value);
+        //Fr::deserialize(c).unwrap()
+        fr
+    }
+
+    fn hash(inputs: &[Self::Fr]) -> Self::Fr {
+        poseidon_hash(inputs)
+    }
+}
+
+// pmtree supports in-memory and on-disk databases (Database trait) for storing the Merkle tree state
+
+// We implement Database for hashmaps, an in-memory database
+struct MemoryDB(HashMap<DBKey, Value>);
+
+impl Database for MemoryDB {
+    fn new(_dbpath: &str) -> Result<Self> {
```

Comment:
As we discussed, maybe each db that implements this trait should have a custom config in the `new` fn. This way we can set db-specific configuration as well.

Reply:
How about a generic?

Reply:
How do you see it ? @rymnc

Reply:
Maybe something like this in pmtree -

(associated type method) -
```rust
trait Database {
  type DbConfig;
  
  fn new(path: &str, config: Self::DbConfig) -> Result<Self>;
  // other funcs
}
```

and then this in rln -
```rust
impl Database for MemoryDB {
  type DbConfig = MemoryDbConfig;
  
  fn new(_dbpath: &str, config: MemoryDbConfig) -> Result<Self> {
    // fn
  }
  
  // other funcs
}
```

OR

(constraints have to be the same across all configs) (generally not used)
(generic trait method) -
in pmtree -
```rust
trait Database<T> {
  fn new(path: &str, config: T) -> Result<Self>;
  // other funcs
}
```

in rln -
```rust
impl Database<MemoryDbConfig> for MemoryDB {
  fn new(_dbpath: &str, config: MemoryDbConfig) -> Result<Self> {
    // fn
  }
  
  // other funcs
}
```

It doesn't make a lot of sense choosing the latter option, because any db would have a concrete type for their config. Let's go with the first option?

Reply:
But what this Config specifies? How it should affect tree implementation, traits, etc. ?

Reply:
This config is db specific. For example, sled db has a flush rate of 500ms. If we wanted to configure it, we can do so here, and by exposing this option in the `new` fn, we can allow users of this lib to set it to whichever value they want. This is just an example, and can evolve with different dbs we use, and how we want to allow the users to configure them.

Reply:
Oh, yes, I got it! That would be good feature, I can add it:)

Reply:
Thank you!😁